### PR TITLE
Improvements to handling of canaries with deployment process deadlines

### PR DESCRIPTION
**Pull request by dadgar**
This PR:
* Allows canary count > desired count of the task group
* Allows rescheduling of canaries and normal allocations while there are canaries placed
* Allows skipping the progress deadline when there are enough healthy canaries and the only waiting change is the promotion.

### CODE REVIEW

File: nomad/deploymentwatcher/deployment_watcher.go [hunk 1]
```diff
@@ -558,36 +564,46 @@ func (w *deploymentWatcher) handleAllocUpdate(allocs []*structs.AllocListStub) (
...
 	}
 
 	d, err := snap.DeploymentByID(nil, w.deploymentID)
 	if err != nil {
-		return false, err
+		return false, false, err
 	}
 
+	fail = false
 	for tg, state := range d.TaskGroups {
```
Comment (preetapan) [inline nomad/deploymentwatcher/deployment_watcher.go]: Is this map always going to be populated with deployment state for all task groups? i.e can some task groups not be in the map initially and then get added later?

File: scheduler/reconcile_util.go [hunk 2]
```diff
@@ -510,21 +506,12 @@ func (a *allocNameIndex) NextCanaries(n uint, existing, destructive allocSet) []
 		}
 	}
 
-	// We have exhausted the preferred and free set, now just pick overlapping
+	// We have exhausted the preferred and free set. Pick starting from n to
+	// n+remainder, to avoid overlapping where possible.
```
Comment (preetapan) [inline scheduler/reconcile_util.go]: Can the comment also have an example, like if desired canaries = 3, and n = 2 it picks indexes 3, 4, 5.. I find this logic hard to follow. Also is there a unit test that asserts the names are what we expect after this specific change?

File: nomad/deploymentwatcher/deployment_watcher.go (hunk 1 above)
Comment (dadgar) [reply nomad/deploymentwatcher/deployment_watcher.go]: No it will only have a mapping for task groups that were being updated at the time of the deployments creation. They are immutable in that sense.

File: scheduler/reconcile_util.go (hunk 2 above)
Comment (dadgar) [reply scheduler/reconcile_util.go]: Here is the unit test: https://github.com/hashicorp/nomad/pull/4206/files#diff-ecc3eb696252654b3d4c86abe54f849fR3152
